lib/pure/optionals.nim

```diff
@@ -0,0 +1,262 @@
+## in a string.
+##
+## .. code-block:: nim
+##
+##   import optionals
+##
+##   proc find(haystack: string, needle: char): ?int =
+##     for i, c in haystack:
+##       if c == needle:
+##         return just i
+##     return nothing(int)  # This line is actually optional,
+##                          # because the default is empty
+##
+## The ``?`` operator (template) is a shortcut for ``Maybe[T]``.
+##
+## .. code-block:: nim
+##
+##   try:
+##     assert("abc".find('c')[] == 2)  # Immediately extract the value
+##   except FieldError:  # If there is no value
+##     assert false  # This will not be reached, because the value is present
+##
+## The ``[]`` operator demonstrated above returns the underlying value, or
+## raises ``FieldError`` if there is no value. There is another option for
+## obtaining the value: ``val``, but you must only use it when you are
+## absolutely sure the value is present (e.g. after checking ``has``). If you do
+## not care about the tiny overhead that ``[]`` causes, you should simply never
+## use ``val``.
+##
+## How to deal with an absence of a value:
+##
+## .. code-block:: nim
+##
+##   let result = "team".find('i')
+##
+##   # Nothing was found, so the result is `nothing`.
+##   assert(result == nothing(int))
+##   # It has no value:
+##   assert(result.has == false)
+##   # A different way to write it:
+##   assert(not result)
+##
+##   try:
+##     echo result[]
+##     assert(false)  # This will not be reached
+##   except FieldError:  # Because an exception is raised
+##     discard
+##
+## Now let's try out the extraction template. It returns whether a value
+## is present and injects the value into a variable. It is meant to be used in
+## a conditional.
+##
+## .. code-block:: nim
+##
+##   if pos ?= "nim".find('i'):
+##     assert(pos is int)  # This is a normal integer, no tricks.
+##     echo "Match found at position ", pos
+##   else:
+##     assert(false)  # This will not be reached
+##
+## Or maybe you want to get the behavior of the standard library's ``find``,
+## which returns `-1` if nothing was found.
+##
+## .. code-block:: nim
+##
+##   assert(("team".find('i') or -1) == -1)
+##   assert(("nim".find('i') or -1) == 1)
+
+import typetraits
+
+
+type
+  Maybe*[T] = object
+    ## An optional type that stores its value and state separately in a boolean.
+    val: T
+    has: bool
+
+
+template `?`*(T: typedesc): typedesc =
+  ## ``?T`` is equivalent to ``Maybe[T]``.
+  Maybe[T]
+
+
+proc just*[T](val: T): Maybe[T] =
+  ## Returns a ``Maybe`` that has this value.
+  result.has = true
+  result.val = val
+
+proc nothing*(T: typedesc): Maybe[T] =
+  ## Returns a ``Maybe`` for this type that has no value.
+  result.has = false
+
+
+proc has*(maybe: Maybe): bool =
+  ## Returns ``true`` if `maybe` isn't `nothing`.
+  maybe.has
+
+converter toBool*(maybe: Maybe): bool =
+  ## Same as ``has``. Allows to use a ``Maybe`` in boolean context.
+  maybe.has
+
+
+proc val*[T](maybe: Maybe[T]): T =
```

Comment:
I agree with flaviut. People who want Maybe[T] also obviously want the extra typing.